### PR TITLE
Fixes for SDLNativeRules issues

### PR DESCRIPTION
**Pull request by ericwol-msft**

### CODE REVIEW

File: src/uws_client.c [hunk 1]
```diff
@@ -1211,7 +1221,7 @@ static void on_underlying_io_bytes_received(void* context, const unsigned char*
 
                     /* Codes_SRS_UWS_CLIENT_01_277: [ To receive WebSocket data, an endpoint listens on the underlying network connection. ]*/
                     /* Codes_SRS_UWS_CLIENT_01_278: [ Incoming data MUST be parsed as WebSocket frames as defined in Section 5.2. ]*/
-                    if (uws_client->stream_buffer_count >= needed_bytes)
+                    if (uws_client->stream_buffer_count >= needed_bytes && uws_client->stream_buffer_size >= 2)
```
Comment (ewertons) [inline src/uws_client.c:1224]: Could you add a comment for why >= 2 is a condition we need to handle?

File: src/uws_client.c [hunk 2]
```diff
@@ -1232,7 +1242,7 @@ static void on_underlying_io_bytes_received(void* context, const unsigned char*
                         {
                             /* Codes_SRS_UWS_CLIENT_01_165: [ If 126, the following 2 bytes interpreted as a 16-bit unsigned integer are the payload length. ]*/
                             needed_bytes += 2;
-                            if (uws_client->stream_buffer_count >= needed_bytes)
+                            if (uws_client->stream_buffer_count >= needed_bytes && uws_client->stream_buffer_size >= 4)
```
Comment (ewertons) [inline src/uws_client.c:1245]: Could you add a comment for why >= 4 is a condition we need to handle?

File: src/uws_client.c [hunk 3]
```diff
@@ -1258,14 +1268,19 @@ static void on_underlying_io_bytes_received(void* context, const unsigned char*
                             needed_bytes += 8;
                             if (uws_client->stream_buffer_count >= needed_bytes)
                             {
-                                if ((uws_client->stream_buffer[2] & 0x80) != 0)
+                                if (uws_client->stream_buffer_size < 10 || uws_client->stream_buffer_size < 2 || (uws_client->stream_buffer[2] & 0x80) != 0)
```
Comment (ewertons) [inline src/uws_client.c:1271]: Could you add a comment for why <10 and <2 are conditions we need to handle?

**Review by ewertons — APPROVED** [review]
Approved with comments.

File: src/string_token.c [hunk 4]
```diff
@@ -22,13 +23,18 @@ typedef struct STRING_TOKEN_TAG
...
+    {
+        LogError("malloc size overflow");
+        result = NULL;
+    }
+    else if ((result = malloc(malloc_size)) == NULL)
     {
         LogError("Failed to allocate array for delimiters lengths");
     }
-    else
+    else if (malloc_size > 0)
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply src/string_token.c:37]: left over from the previous overflow check, removed

File: src/buffer.c [hunk 5]
```diff
@@ -540,20 +541,21 @@ int BUFFER_prepend(BUFFER_HANDLE handle1, BUFFER_HANDLE handle2)
         else
         {
             //put b2 ahead of b1: [b2][b1], return b1
+            size_t malloc_size = safe_add_size_t(b1->size, b2->size);
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply src/buffer.c:544]: correct SIZE_MAX is being used for an overflow flag.

File: src/http_proxy_io.c [hunk 6]
```diff
@@ -631,8 +632,16 @@ static void on_underlying_io_bytes_received(void* context, const unsigned char*
         case HTTP_PROXY_IO_STATE_WAITING_FOR_CONNECT_RESPONSE:
         {
             /* Codes_SRS_HTTP_PROXY_IO_01_065: [ When bytes are received and the response to the CONNECT request was not yet received, the bytes shall be accumulated until a double new-line is detected. ]*/
-            unsigned char* new_receive_buffer = (unsigned char*)realloc(http_proxy_io_instance->receive_buffer, http_proxy_io_instance->receive_buffer_size + size + 1);
-            if (new_receive_buffer == NULL)
+            // size_t malloc_size = http_proxy_io_instance->receive_buffer_size + size + 1;
+            size_t realloc_size = safe_add_size_t(safe_add_size_t(http_proxy_io_instance->receive_buffer_size, size), 1);
+
+            unsigned char* new_receive_buffer = NULL;
+            if (realloc_size == SIZE_MAX)
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply src/http_proxy_io.c:639]: correct SIZE_MAX is being used for an overflow flag.

File: src/strings.c [hunk 7]
```diff
@@ -297,12 +298,23 @@ STRING_HANDLE STRING_new_JSON(const char* source)
...
         else
         {
-            if ((result = (STRING*)malloc(sizeof(STRING))) == NULL)
+            //size_t malloc_len = vlen + 5 * nControlCharacters + nEscapeCharacters + 3;
+            size_t malloc_len = safe_multiply_size_t(5, nControlCharacters);
+            malloc_len = safe_add_size_t(malloc_len, vlen);
+            malloc_len = safe_add_size_t(malloc_len, nEscapeCharacters);
+            malloc_len = safe_add_size_t(malloc_len, 3);
+
+            if (malloc_len == SIZE_MAX)
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply src/strings.c:307]: correct SIZE_MAX is being used for an overflow flag.

File: src/string_token.c [hunk 8]
```diff
@@ -41,7 +47,7 @@ static size_t* get_delimiters_lengths(const char** delimiters, size_t n_delims)
                 result = NULL;
                 break;
             }
-            else
+            else if (((i+1) * sizeof(size_t)) < malloc_size)
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply src/string_token.c]: SDL scan is not smart that result is an array.

File: src/strings.c [hunk 9]
```diff
@@ -297,12 +298,23 @@ STRING_HANDLE STRING_new_JSON(const char* source)
         }
         else
         {
-            if ((result = (STRING*)malloc(sizeof(STRING))) == NULL)
+            //size_t malloc_len = vlen + 5 * nControlCharacters + nEscapeCharacters + 3;
```
Comment (ewertons) [inline src/strings.c:301]: Should remove?

File: src/uws_client.c [hunk 10]
```diff
@@ -1093,8 +1095,12 @@ static void on_underlying_io_bytes_received(void* context, const unsigned char*
             case UWS_STATE_WAITING_FOR_UPGRADE_RESPONSE:
             {
                 /* Codes_SRS_UWS_CLIENT_01_378: [ When on_underlying_io_bytes_received is called while the uws is OPENING, the received bytes shall be accumulated in order to attempt parsing the WebSocket Upgrade response. ]*/
-                unsigned char* new_received_bytes = (unsigned char*)realloc(uws_client->stream_buffer, uws_client->stream_buffer_count + size + 1);
-                if (new_received_bytes == NULL)
+                unsigned char* new_received_bytes;
+
+                //size_t realloc_size = uws_client->stream_buffer_count + size + 1; **using safe int**
```
Comment (ewertons) [inline src/uws_client.c:1100]: Should remove?

File: src/uws_client.c [hunk 11]
```diff
@@ -1116,8 +1122,12 @@ static void on_underlying_io_bytes_received(void* context, const unsigned char*
             case UWS_STATE_CLOSING_WAITING_FOR_CLOSE:
             {
                 /* Codes_SRS_UWS_CLIENT_01_385: [ If the state of the uws instance is OPEN, the received bytes shall be used for decoding WebSocket frames. ]*/
-                unsigned char* new_received_bytes = (unsigned char*)realloc(uws_client->stream_buffer, uws_client->stream_buffer_count + size + 1);
-                if (new_received_bytes == NULL)
+                unsigned char* new_received_bytes;
+
+                //size_t realloc_size = uws_client->stream_buffer_count + size + 1; **using safe int**
```
Comment (ewertons) [inline src/uws_client.c:1127]: Should remove?

**Review by ewertons — APPROVED** [review]

File: src/strings.c (hunk 9 above)
Comment (ericwol-msft) [reply src/strings.c:301]: No, I want it there to show the intended math using safe macros

File: src/uws_client.c (hunk 11 above)
Comment (ericwol-msft) [reply src/uws_client.c:1127]: No, I want it there to show the intended math using safe macros

File: src/uws_client.c (hunk 10 above)
Comment (ericwol-msft) [reply src/uws_client.c:1100]: No, I want it there to show the intended math using safe macros

**Pull request merged**